File: apps/iris/src-tauri/plugins/mobile-bluetooth/android/src/main/java/to/iris/browser/mobilebluetooth/MobileBluetoothPlugin.kt
package to.iris.browser.mobilebluetooth

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothGattServer
import android.bluetooth.BluetoothGattServerCallback
import android.bluetooth.BluetoothGattService
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothStatusCodes
import android.bluetooth.le.AdvertiseCallback
import android.bluetooth.le.AdvertiseData
import android.bluetooth.le.AdvertiseSettings
import android.bluetooth.le.BluetoothLeAdvertiser
import android.content.Context
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.ParcelUuid
import android.util.Base64
import android.util.Log
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import org.json.JSONArray

private const val TAG = "MobileBluetoothPlugin"
private val SERVICE_UUID: UUID = UUID.fromString("f18ef5f6-b7ee-4f40-b869-10a2d4f35932")
private val RX_UUID: UUID = UUID.fromString("0bb5f5c9-6369-4511-a84f-4d4c14d8f8d4")
private val TX_UUID: UUID = UUID.fromString("4ec9c0c2-97c6-4f46-9fd1-927d699b2f6d")
private val CCCD_UUID: UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
private const val CHUNK_BYTES: Int = 180
private const val UNREADY_DISCONNECT_DELAY_MS: Long = 2500

@InvokeArg
class StartArgs {
    lateinit var localPeerId: String
}

@InvokeArg
class SendArgs {
    lateinit var address: String
    lateinit var kind: String
    lateinit var payloadBase64: String
}

private data class DecodedFrame(val kind: String, val payload: ByteArray)

private data class QueuedFrame(val address: String, val kind: String, val payloadBase64: String)

private class FrameDecoder {
    private val buffer = ArrayList<Byte>()

    fun append(chunk: ByteArray): List<DecodedFrame> {
        val frames = mutableListOf<DecodedFrame>()
        chunk.forEach { buffer.add(it) }
        while (buffer.size >= 5) {
            val header = ByteBuffer.wrap(byteArrayOf(buffer[1], buffer[2], buffer[3], buffer[4]))
            val len = header.int
            if (buffer.size < 5 + len) {
                break
            }
            val kind = buffer[0].toInt()
            val payload = ByteArray(len)
            for (i in 0 until len) {
                payload[i] = buffer[5 + i]
            }
            repeat(5 + len) { buffer.removeAt(0) }
            frames.add(
                DecodedFrame(
                    if (kind == 1) "text" else "binary",
                    payload,
                )
            )
        }
        return frames
    }
}

private fun encodeFrame(kind: String, payload: ByteArray): ByteArray {
    val kindByte: Byte = if (kind == "text") 1 else 2
    val header = ByteBuffer.allocate(5)
    header.put(kindByte)
    header.putInt(payload.size)
    return header.array() + payload
}

private fun helloPayload(localPeerId: String): ByteArray {
    return """{"type":"hello","peerId":"$localPeerId"}""".toByteArray(StandardCharsets.UTF_8)
}

@TauriPlugin
class MobileBluetoothPlugin(private val activity: android.app.Activity) : Plugin(activity) {
    private val appContext: Context = activity.applicationContext
    private val bluetoothManager =
        appContext.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
    private val bluetoothAdapter: BluetoothAdapter? = bluetoothManager.adapter
    private val advertiser: BluetoothLeAdvertiser? = bluetoothAdapter?.bluetoothLeAdvertiser

    private var gattServer: BluetoothGattServer? = null
    private var txCharacteristic: BluetoothGattCharacteristic? = null
    private var localPeerId: String = ""
    private val devices = ConcurrentHashMap<String, BluetoothDevice>()
    private val subscribed = ConcurrentHashMap.newKeySet<String>()
    private val decoders = ConcurrentHashMap<String, FrameDecoder>()
    private val drainedFrames = ConcurrentLinkedQueue<QueuedFrame>()
    private val pendingDisconnects = ConcurrentHashMap<String, Runnable>()
    private val mainHandler = Handler(Looper.getMainLooper())
    private val advertiseCallback = object : AdvertiseCallback() {}
    private var bluetoothActive = false

    @Command
    fun start(invoke: Invoke) {
        val args = invoke.parseArgs(StartArgs::class.java)
        localPeerId = args.localPeerId
        Log.i(TAG, "start invoked for peer ${args.localPeerId}")
        try {
            stopInternal()
            if (bluetoothAdapter == null || advertiser == null) {
                Log.w(TAG, "Bluetooth LE advertiser is unavailable")
                invoke.reject("Bluetooth LE advertiser is unavailable")
                return
            }
            if (!bluetoothAdapter.isEnabled) {
                Log.w(TAG, "Bluetooth is disabled")
                invoke.reject("Bluetooth is disabled")
                return
            }

            val serviceAdded = CountDownLatch(1)
            var serviceAddedOk = false

            val serverCallback = object : BluetoothGattServerCallback() {
                override fun onConnectionStateChange(device: BluetoothDevice, status: Int, newState: Int) {
                    val address = device.address
                    if (newState == android.bluetooth.BluetoothProfile.STATE_CONNECTED) {
                        Log.d(TAG, "BLE device connected: $address")
                        devices[address] = device
                        decoders[address] = FrameDecoder()
                        scheduleUnreadyDisconnect(address, device, "startup-timeout")
                        triggerAddress("peer-connected", address)
                    } else if (newState == android.bluetooth.BluetoothProfile.STATE_DISCONNECTED) {
                        Log.d(TAG, "BLE device disconnected: $address")
                        cancelPendingDisconnect(address)
                        devices.remove(address)
                        subscribed.remove(address)
                        decoders.remove(address)
                        triggerAddress("peer-disconnected", address)
                    }
                }

                override fun onCharacteristicReadRequest(
                    device: BluetoothDevice,
                    requestId: Int,
                    offset: Int,
                    characteristic: BluetoothGattCharacteristic,
                ) {
                    if (characteristic.uuid == TX_UUID) {
                        val frame = encodeFrame("text", helloPayload(localPeerId))
                        val value =
                            if (offset in 0 until frame.size) frame.copyOfRange(offset, frame.size) else ByteArray(0)
                        Log.d(TAG, "Serving hello read to ${device.address} (${value.size} bytes, offset=$offset)")
                        gattServer?.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, offset, value)
                    } else {
                        gattServer?.sendResponse(device, requestId, BluetoothGatt.GATT_FAILURE, offset, null)
                    }
                }

                override fun onCharacteristicWriteRequest(
                    device: BluetoothDevice,
                    requestId: Int,
                    characteristic: BluetoothGattCharacteristic,
                    preparedWrite: Boolean,
                    responseNeeded: Boolean,
                    offset: Int,
                    value: ByteArray,
                ) {
                    if (characteristic.uuid == RX_UUID) {
                        Log.d(TAG, "Received BLE write from ${device.address} (${value.size} bytes)")
                        val decoder = decoders.computeIfAbsent(device.address) { FrameDecoder() }
                        decoder.append(value).forEach { frame ->
                            triggerFrame(device.address, frame.kind, frame.payload)
                        }
                    }
                    if (responseNeeded) {
                        gattServer?.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, 0, null)
                    }
                }

                override fun onDescriptorWriteRequest(
                    device: BluetoothDevice,
                    requestId: Int,
                    descriptor: BluetoothGattDescriptor,
                    preparedWrite: Boolean,
                    responseNeeded: Boolean,
                    offset: Int,
                    value: ByteArray,
                ) {
                    if (descriptor.uuid == CCCD_UUID && value.contentEquals(BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE)) {
                        Log.d(TAG, "Notifications enabled for ${device.address}")
                        subscribed.add(device.address)
                        cancelPendingDisconnect(device.address)
                        sendHello(device)
                        triggerAddress("peer-ready", device.address)
                    }
                    if (responseNeeded) {
                        gattServer?.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, 0, null)
                    }
                }

                override fun onServiceAdded(status: Int, service: BluetoothGattService) {
                    if (service.uuid != SERVICE_UUID) {
                        return
                    }
                    serviceAddedOk = status == BluetoothGatt.GATT_SUCCESS
                    Log.i(TAG, "Service add callback for ${service.uuid} status=$status")
                    serviceAdded.countDown()
                }
            }

            gattServer = bluetoothManager.openGattServer(appContext, serverCallback)
            if (gattServer == null) {
                Log.e(TAG, "Failed to open Bluetooth GATT server")
                invoke.reject("Failed to open Bluetooth GATT server")
                return
            }
            val rx = BluetoothGattCharacteristic(
                RX_UUID,
                BluetoothGattCharacteristic.PROPERTY_WRITE or BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE,
                BluetoothGattCharacteristic.PERMISSION_WRITE,
            )
            val tx = BluetoothGattCharacteristic(
                TX_UUID,
                BluetoothGattCharacteristic.PROPERTY_NOTIFY or BluetoothGattCharacteristic.PROPERTY_READ,
                BluetoothGattCharacteristic.PERMISSION_READ,
            )
            tx.addDescriptor(
                BluetoothGattDescriptor(
                    CCCD_UUID,
                    BluetoothGattDescriptor.PERMISSION_READ or BluetoothGattDescriptor.PERMISSION_WRITE,
                )
            )
            txCharacteristic = tx
            val service = BluetoothGattService(SERVICE_UUID, BluetoothGattService.SERVICE_TYPE_PRIMARY)
            service.addCharacteristic(rx)
            service.addCharacteristic(tx)
            if (gattServer?.addService(service) != true) {
                Log.e(TAG, "Failed to queue Bluetooth GATT service")
                invoke.reject("Failed to add Bluetooth GATT service")
                return
            }
            if (!serviceAdded.await(3, TimeUnit.SECONDS)) {
                Log.e(TAG, "Timed out waiting for Bluetooth GATT service registration")
                invoke.reject("Timed out waiting for Bluetooth GATT service registration")
                return
            }
            if (!serviceAddedOk) {
                Log.e(TAG, "Bluetooth GATT service registration failed")
                invoke.reject("Bluetooth GATT service registration failed")
                return
            }
            Log.i(TAG, "Bluetooth GATT service registered, starting advertising")

            advertiser.startAdvertising(
                AdvertiseSettings.Builder()
                    .setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_LOW_LATENCY)
                    .setTxPowerLevel(AdvertiseSettings.ADVERTISE_TX_POWER_HIGH)
                    .setConnectable(true)
                    .build(),
                AdvertiseData.Builder()
                    .addServiceUuid(ParcelUuid(SERVICE_UUID))
                    .setIncludeDeviceName(false)
                    .build(),
                advertiseCallback,
            )

            bluetoothActive = true
            Log.i(TAG, "Bluetooth advertising and GATT server started")
            invoke.resolve()
        } catch (error: SecurityException) {
            Log.e(TAG, "Bluetooth permission missing", error)
            invoke.reject("Bluetooth permission missing: ${error.message}")
        } catch (error: Exception) {
            Log.e(TAG, "Failed to start Bluetooth server", error)
            invoke.reject("Failed to start Bluetooth server: ${error.message}")
        }
    }

    @Command
    fun stop(invoke: Invoke) {
        Log.i(TAG, "stop invoked")
        bluetoothActive = false
        stopInternal()
        invoke.resolve()
    }

    @Command
    fun send(invoke: Invoke) {
        val args = invoke.parseArgs(SendArgs::class.java)
        val device = devices[args.address]
        val characteristic = txCharacteristic
        val server = gattServer
        if (device == null || characteristic == null || server == null) {
            invoke.reject("Bluetooth peer not connected")
            return
        }
        if (!subscribed.contains(args.address)) {
            invoke.reject("Bluetooth peer is not ready for notifications")
            return
        }
        val payload = Base64.decode(args.payloadBase64, Base64.DEFAULT)
        val frame = encodeFrame(args.kind, payload)
        for (chunk in frame.asList().chunked(CHUNK_BYTES)) {
            if (!notifyCharacteristicChanged(server, device, characteristic, chunk.toByteArray())) {
                invoke.reject("Failed to notify Bluetooth peer")
                return
            }
        }
        invoke.resolve()
    }

    @Command
    fun pollTransport(invoke: Invoke) {
        Log.i(TAG, "pollTransport returning ${devices.size} device(s); ready=${subscribed.size}")
        invoke.resolve(buildTransportPollPayload())
    }

    override fun onDestroy() {
        super.onDestroy()
        bluetoothActive = false
        stopInternal()
    }

    private fun sendHello(device: BluetoothDevice) {
        val characteristic = txCharacteristic ?: return
        val server = gattServer ?: return
        val frame = encodeFrame("text", helloPayload(localPeerId))
        for (chunk in frame.asList().chunked(CHUNK_BYTES)) {
            notifyCharacteristicChanged(server, device, characteristic, chunk.toByteArray())
        }
    }

    private fun triggerAddress(event: String, address: String) {
        val payload = JSObject()
        payload.put("address", address)
        trigger(event, payload)
    }

    private fun triggerFrame(address: String, kind: String, payloadBytes: ByteArray) {
        val payloadBase64 = Base64.encodeToString(payloadBytes, Base64.NO_WRAP)
        drainedFrames.add(QueuedFrame(address, kind, payloadBase64))
    }

    private fun stopInternal() {
        Log.d(TAG, "Stopping Bluetooth advertiser and GATT server")
        disconnectUnreadyDevices("stop")
        try {
            advertiser?.stopAdvertising(advertiseCallback)
        } catch (_: Exception) {}
        try {
            gattServer?.close()
        } catch (_: Exception) {}
        gattServer = null
        txCharacteristic = null
        devices.clear()
        subscribed.clear()
        decoders.clear()
        drainedFrames.clear()
        pendingDisconnects.values.forEach { mainHandler.removeCallbacks(it) }
        pendingDisconnects.clear()
    }

    private fun buildTransportPollPayload(): JSObject {
        val payload = JSObject()
        payload.put("peers", buildPeerSnapshots())
        payload.put("frames", drainQueuedFrames())
        return payload
    }

    private fun buildPeerSnapshots(): JSONArray {
        val peers = JSONArray()
        devices.keys.sorted().forEach { address ->
            val peer = JSObject()
            peer.put("address", address)
            peer.put("ready", subscribed.contains(address))
            peers.put(peer)
        }
        return peers
    }

    private fun drainQueuedFrames(): JSONArray {
        val frames = JSONArray()
        while (true) {
            val frame = drainedFrames.poll() ?: break
            val payload = JSObject()
            payload.put("address", frame.address)
            payload.put("kind", frame.kind)
            payload.put("payloadBase64", frame.payloadBase64)
            frames.put(payload)
        }
        return frames
    }

    private fun disconnectUnreadyDevices(reason: String) {
        val server = gattServer ?: return
        devices.entries
            .sortedBy { it.key }
            .forEach { (address, device) ->
                if (subscribed.contains(address)) {
                    return@forEach
                }
                Log.i(TAG, "Disconnecting BLE device $address to recover clean startup ($reason)")
                try {
                    server.cancelConnection(device)
                } catch (error: Exception) {
                    Log.w(TAG, "Failed to cancel BLE connection for $address", error)
                }
            }
    }

    private fun scheduleUnreadyDisconnect(address: String, device: BluetoothDevice, reason: String) {
        cancelPendingDisconnect(address)
        val disconnectTask =
            Runnable {
                if (!bluetoothActive || subscribed.contains(address)) {
                    pendingDisconnects.remove(address)
                    return@Runnable
                }
                Log.i(TAG, "Disconnecting BLE device $address after readiness timeout ($reason)")
                try {
                    gattServer?.cancelConnection(device)
                } catch (error: Exception) {
                    Log.w(TAG, "Failed to cancel BLE connection for $address", error)
                } finally {
                    pendingDisconnects.remove(address)
                }
            }
        pendingDisconnects[address] = disconnectTask
        mainHandler.postDelayed(disconnectTask, UNREADY_DISCONNECT_DELAY_MS)
    }

    private fun cancelPendingDisconnect(address: String) {
        pendingDisconnects.remove(address)?.let { pending ->
            mainHandler.removeCallbacks(pending)
        }
    }

    @Suppress("DEPRECATION")
    private fun notifyCharacteristicChanged(
        server: BluetoothGattServer,
        device: BluetoothDevice,
        characteristic: BluetoothGattCharacteristic,
        value: ByteArray,
    ): Boolean {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            server.notifyCharacteristicChanged(device, characteristic, false, value) ==
                BluetoothStatusCodes.SUCCESS
        } else {
            characteristic.value = value
            server.notifyCharacteristicChanged(device, characteristic, false)
        }
    }
}

File: apps/iris/src-tauri/plugins/mobile-bluetooth/ios/Sources/MobileBluetoothPlugin.swift
import CoreBluetooth
import Foundation
import Tauri

struct StartArgs: Decodable {
  let localPeerId: String
}

struct SendArgs: Decodable {
  let address: String
  let kind: String
  let payloadBase64: String
}

private struct AddressEvent: Encodable {
  let address: String
}

private struct FrameEvent: Encodable {
  let address: String
  let kind: String
  let payloadBase64: String
}

private struct PeerSnapshot: Encodable {
  let address: String
  let ready: Bool
}

private struct PeerSnapshotResponse: Encodable {
  let peers: [PeerSnapshot]
}

private struct DecodedFrame {
  let kind: String
  let payload: Data
}

private struct PendingSend {
  let address: String
  let chunks: [Data]
  var nextIndex: Int
  let invoke: Invoke?
}

private let serviceUUID = CBUUID(string: "f18ef5f6-b7ee-4f40-b869-10a2d4f35932")
private let rxUUID = CBUUID(string: "0bb5f5c9-6369-4511-a84f-4d4c14d8f8d4")
private let txUUID = CBUUID(string: "4ec9c0c2-97c6-4f46-9fd1-927d699b2f6d")
private let chunkBytes = 180

private final class FrameDecoder {
  private var buffer = [UInt8]()

  func append(_ chunk: Data) -> [DecodedFrame] {
    buffer.append(contentsOf: chunk)
    var frames = [DecodedFrame]()

    while buffer.count >= 5 {
      let length =
        (Int(buffer[1]) << 24) | (Int(buffer[2]) << 16) | (Int(buffer[3]) << 8) | Int(buffer[4])
      guard buffer.count >= 5 + length else {
        break
      }

      let payload = Data(buffer[5..<(5 + length)])
      let kind: String
      switch buffer[0] {
      case 1:
        kind = "text"
      case 2:
        kind = "binary"
      default:
        buffer.removeAll()
        return frames
      }

      frames.append(DecodedFrame(kind: kind, payload: payload))
      buffer.removeFirst(5 + length)
    }

    return frames
  }
}

private func encodeFrame(kind: String, payload: Data) -> Data? {
  let kindByte: UInt8
  switch kind {
  case "text":
    kindByte = 1
  case "binary":
    kindByte = 2
  default:
    return nil
  }

  let length = UInt32(payload.count).bigEndian
  var header = Data([kindByte])
  withUnsafeBytes(of: length) { bytes in
    header.append(contentsOf: bytes)
  }
  var frame = header
  frame.append(payload)
  return frame
}

private func helloPayload(localPeerId: String) -> Data {
  Data(#"{"type":"hello","peerId":"\#(localPeerId)"}"#.utf8)
}

private extension Data {
  func chunked(maxLength: Int) -> [Data] {
    guard !isEmpty else {
      return [Data()]
    }

    var chunks = [Data]()
    var index = startIndex
    while index < endIndex {
      let nextIndex = self.index(index, offsetBy: maxLength, limitedBy: endIndex) ?? endIndex
      chunks.append(self[index..<nextIndex])
      index = nextIndex
    }
    return chunks
  }
}

class MobileBluetoothPlugin: Plugin, CBPeripheralManagerDelegate {
  private var peripheralManager: CBPeripheralManager?
  private var rxCharacteristic: CBMutableCharacteristic?
  private var txCharacteristic: CBMutableCharacteristic?
  private var localPeerId = ""
  private var desiredActive = false
  private var bluetoothActive = false
  private var serviceRegistrationPending = false
  private var advertisingPending = false
  private var peers = [String: CBCentral]()
  private var readyPeers = Set<String>()
  private var decoders = [String: FrameDecoder]()
  private var pendingSends = [PendingSend]()
  private var pendingStartInvoke: Invoke?

  @objc public func start(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(StartArgs.self)

    DispatchQueue.main.async {
      self.pendingStartInvoke?.reject("Bluetooth start superseded by a newer request")
      self.pendingStartInvoke = invoke
      self.localPeerId = args.localPeerId
      self.desiredActive = true
      self.resetTransportState(rejectPendingSends: true)
      self.ensurePeripheralManager()
      self.maybeStartPeripheral()
    }
  }

  @objc public func stop(_ invoke: Invoke) throws {
    DispatchQueue.main.async {
      self.desiredActive = false
      self.failPendingStart("Bluetooth stopped")
      self.resetTransportState(rejectPendingSends: true)
      invoke.resolve()
    }
  }

  @objc public func send(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(SendArgs.self)

    DispatchQueue.main.async {
      guard self.bluetoothActive else {
        invoke.reject("Bluetooth peer not connected")
        return
      }
      guard let payload = Data(base64Encoded: args.payloadBase64) else {
        invoke.reject("Invalid Bluetooth payload")
        return
      }
      self.queueFrame(to: args.address, kind: args.kind, payload: payload, invoke: invoke)
    }
  }

  @objc public func listPeers(_ invoke: Invoke) throws {
    DispatchQueue.main.async {
      let peers = self.peers.keys.sorted().map { address in
        PeerSnapshot(address: address, ready: self.readyPeers.contains(address))
      }
      invoke.resolve(PeerSnapshotResponse(peers: peers))
    }
  }

  func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
    DispatchQueue.main.async {
      switch peripheral.state {
      case .poweredOn:
        self.maybeStartPeripheral()
      case .poweredOff:
        self.failActiveSession("Bluetooth is disabled")
      case .unauthorized:
        self.failActiveSession("Bluetooth permission denied")
      case .unsupported:
        self.failActiveSession("Bluetooth LE peripheral is unavailable")
      case .resetting, .unknown:
        self.bluetoothActive = false
      @unknown default:
        self.failActiveSession("Bluetooth state is unavailable")
      }
    }
  }

  func peripheralManager(_ peripheral: CBPeripheralManager, didAdd service: CBService, error: Error?) {
    DispatchQueue.main.async {
      guard service.uuid == serviceUUID else {
        return
      }
      self.serviceRegistrationPending = false
      if let error = error {
        self.failActiveSession("Failed to add Bluetooth GATT service: \(error.localizedDescription)")
        return
      }

      self.advertisingPending = true
      peripheral.startAdvertising([
        CBAdvertisementDataServiceUUIDsKey: [serviceUUID]
      ])
    }
  }

  func peripheralManagerDidStartAdvertising(_ peripheral: CBPeripheralManager, error: Error?) {
    DispatchQueue.main.async {
      self.advertisingPending = false
      if let error = error {
        self.failActiveSession("Failed to start Bluetooth advertising: \(error.localizedDescription)")
        return
      }

      self.bluetoothActive = true
      self.pendingStartInvoke?.resolve()
      self.pendingStartInvoke = nil
    }
  }

  func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveRead request: CBATTRequest) {
    DispatchQueue.main.async {
      guard request.characteristic.uuid == txUUID else {
        peripheral.respond(to: request, withResult: .requestNotSupported)
        return
      }

      _ = self.rememberPeer(request.central)
      guard let frame = encodeFrame(kind: "text", payload: helloPayload(localPeerId: self.localPeerId)) else {
        peripheral.respond(to: request, withResult: .unlikelyError)
        return
      }
      guard request.offset <= frame.count else {
        peripheral.respond(to: request, withResult: .invalidOffset)
        return
      }

      request.value = frame.subdata(in: request.offset..<frame.count)
      peripheral.respond(to: request, withResult: .success)
    }
  }

  func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveWrite requests: [CBATTRequest]) {
    DispatchQueue.main.async {
      guard let firstRequest = requests.first else {
        return
      }

      var result: CBATTError.Code = .success
      for request in requests {
        guard request.characteristic.uuid == rxUUID else {
          result = .requestNotSupported
          break
        }

        let address = self.rememberPeer(request.central)
        guard let value = request.value else {
          continue
        }

        let decoder = self.decoders[address] ?? FrameDecoder()
        self.decoders[address] = decoder
        for frame in decoder.append(value) {
          self.triggerFrame(address: address, kind: frame.kind, payload: frame.payload)
        }
      }

      peripheral.respond(to: firstRequest, withResult: result)
    }
  }

  func peripheralManager(
    _ peripheral: CBPeripheralManager,
    central: CBCentral,
    didSubscribeTo characteristic: CBCharacteristic
  ) {
    DispatchQueue.main.async {
      guard characteristic.uuid == txUUID else {
        return
      }

      let address = self.rememberPeer(central)
      self.readyPeers.insert(address)
      self.sendHello(to: address)
      self.triggerAddress("peer-ready", address: address)
    }
  }

  func peripheralManager(
    _ peripheral: CBPeripheralManager,
    central: CBCentral,
    didUnsubscribeFrom characteristic: CBCharacteristic
  ) {
    DispatchQueue.main.async {
      guard characteristic.uuid == txUUID else {
        return
      }

      self.removePeer(address: self.address(for: central), notify: true)
    }
  }

  func peripheralManagerIsReady(toUpdateSubscribers peripheral: CBPeripheralManager) {
    DispatchQueue.main.async {
      self.flushPendingSends()
    }
  }

  private func ensurePeripheralManager() {
    if peripheralManager == nil {
      peripheralManager = CBPeripheralManager(
        delegate: self,
        queue: nil,
        options: [CBPeripheralManagerOptionShowPowerAlertKey: true]
      )
    } else {
      peripheralManager?.delegate = self
    }
  }

  private func maybeStartPeripheral() {
    guard desiredActive, let peripheral = peripheralManager else {
      return
    }

    switch peripheral.state {
    case .poweredOn:
      guard !bluetoothActive, !serviceRegistrationPending, !advertisingPending else {
        return
      }

      let rx = CBMutableCharacteristic(
        type: rxUUID,
        properties: [.write, .writeWithoutResponse],
        value: nil,
        permissions: [.writeable]
      )
      let tx = CBMutableCharacteristic(
        type: txUUID,
        properties: [.notify, .read],
        value: nil,
        permissions: [.readable]
      )

      let service = CBMutableService(type: serviceUUID, primary: true)
      service.characteristics = [rx, tx]

      rxCharacteristic = rx
      txCharacteristic = tx
      serviceRegistrationPending = true
      peripheral.add(service)
    case .poweredOff:
      failActiveSession("Bluetooth is disabled")
    case .unauthorized:
      failActiveSession("Bluetooth permission denied")
    case .unsupported:
      failActiveSession("Bluetooth LE peripheral is unavailable")
    case .resetting, .unknown:
      break
    @unknown default:
      failActiveSession("Bluetooth state is unavailable")
    }
  }

  private func failActiveSession(_ message: String) {
    let wasActive = desiredActive || bluetoothActive || serviceRegistrationPending || advertisingPending
    desiredActive = false
    failPendingStart(message)
    if wasActive {
      resetTransportState(rejectPendingSends: true)
    }
  }

  private func failPendingStart(_ message: String) {
    pendingStartInvoke?.reject(message)
    pendingStartInvoke = nil
  }

  private func resetTransportState(rejectPendingSends: Bool) {
    peripheralManager?.stopAdvertising()
    peripheralManager?.removeAllServices()
    bluetoothActive = false
    serviceRegistrationPending = false
    advertisingPending = false
    rxCharacteristic = nil
    txCharacteristic = nil
    peers.removeAll()
    readyPeers.removeAll()
    decoders.removeAll()

    if rejectPendingSends {
      rejectAllPendingSends("Bluetooth stopped")
    } else {
      pendingSends.removeAll()
    }
  }

  private func queueFrame(to address: String, kind: String, payload: Data, invoke: Invoke?) {
    guard readyPeers.contains(address), peers[address] != nil else {
      invoke?.reject("Bluetooth peer is not ready for notifications")
      return
    }
    guard let frame = encodeFrame(kind: kind, payload: payload) else {
      invoke?.reject("Invalid Bluetooth frame kind")
      return
    }

    pendingSends.append(
      PendingSend(address: address, chunks: frame.chunked(maxLength: chunkBytes), nextIndex: 0, invoke: invoke)
    )
    flushPendingSends()
  }

  private func flushPendingSends() {
    guard let peripheral = peripheralManager, let txCharacteristic else {
      return
    }

    while !pendingSends.isEmpty {
      var send = pendingSends.removeFirst()
      guard readyPeers.contains(send.address), let central = peers[send.address] else {
        send.invoke?.reject("Bluetooth peer not connected")
        continue
      }

      while send.nextIndex < send.chunks.count {
        let chunk = send.chunks[send.nextIndex]
        if peripheral.updateValue(chunk, for: txCharacteristic, onSubscribedCentrals: [central]) {
          send.nextIndex += 1
        } else {
          pendingSends.insert(send, at: 0)
          return
        }
      }

      send.invoke?.resolve()
    }
  }

  private func sendHello(to address: String) {
    queueFrame(to: address, kind: "text", payload: helloPayload(localPeerId: localPeerId), invoke: nil)
  }

  private func rememberPeer(_ central: CBCentral) -> String {
    let address = self.address(for: central)
    if peers[address] == nil {
      peers[address] = central
      decoders[address] = FrameDecoder()
      triggerAddress("peer-connected", address: address)
    } else {
      peers[address] = central
    }
    return address
  }

  private func removePeer(address: String, notify: Bool) {
    peers.removeValue(forKey: address)
    readyPeers.remove(address)
    decoders.removeValue(forKey: address)
    rejectPendingSends(for: address, message: "Bluetooth peer disconnected")
    if notify {
      triggerAddress("peer-disconnected", address: address)
    }
  }

  private func rejectPendingSends(for address: String, message: String) {
    var retained = [PendingSend]()
    for send in pendingSends {
      if send.address == address {
        send.invoke?.reject(message)
      } else {
        retained.append(send)
      }
    }
    pendingSends = retained
  }

  private func rejectAllPendingSends(_ message: String) {
    for send in pendingSends {
      send.invoke?.reject(message)
    }
    pendingSends.removeAll()
  }

  private func address(for central: CBCentral) -> String {
    central.identifier.uuidString
  }

  private func triggerAddress(_ event: String, address: String) {
    try? trigger(event, data: AddressEvent(address: address))
  }

  private func triggerFrame(address: String, kind: String, payload: Data) {
    try? trigger(
      "frame",
      data: FrameEvent(address: address, kind: kind, payloadBase64: payload.base64EncodedString())
    )
  }
}

@_cdecl("init_plugin_mobile_bluetooth")
func initPlugin() -> Plugin {
  MobileBluetoothPlugin()
}
